tracer/src/Datadog.Trace/Propagators/SpanContextPropagator.cs

```diff
@@ -150,76 +154,101 @@ internal void Inject<TCarrier, TCarrierSetter>(SpanContext context, TCarrier car
         /// <param name="getter">The function that can extract a list of values for a given header name.</param>
         /// <typeparam name="TCarrier">Type of header collection</typeparam>
         /// <returns>A new <see cref="SpanContext"/> that contains the values obtained from <paramref name="carrier"/>.</returns>
-        public SpanContext? Extract<TCarrier>(TCarrier carrier, Func<TCarrier, string, IEnumerable<string?>> getter)
+        public PropagationContext Extract<TCarrier>(TCarrier carrier, Func<TCarrier, string, IEnumerable<string?>> getter)
         {
             if (carrier == null) { ThrowHelper.ThrowArgumentNullException(nameof(carrier)); }
             if (getter == null!) { ThrowHelper.ThrowArgumentNullException(nameof(getter)); }
 
             return Extract(carrier, new FuncGetter<TCarrier>(getter));
         }
 
-        internal SpanContext? Extract<TCarrier, TCarrierGetter>(TCarrier carrier, TCarrierGetter carrierGetter)
+        internal PropagationContext Extract<TCarrier, TCarrierGetter>(TCarrier carrier, TCarrierGetter carrierGetter)
             where TCarrierGetter : struct, ICarrierGetter<TCarrier>
         {
             if (carrier is null) { ThrowHelper.ThrowArgumentNullException(nameof(carrier)); }
 
-            SpanContext? localSpanContext = null;
+            // as we extract values from the carrier using multiple extractors,
+            // we will accumulate them in this context
+            SpanContext? cumulativeSpanContext = null;
+            Baggage? cumulativeBaggage = null;
 
-            for (var i = 0; i < _extractors.Length; i++)
+            foreach (var extractor in _extractors)
             {
-                if (_extractors[i].TryExtract(carrier, carrierGetter, out var spanContext))
+                if (_propagationExtractFirstOnly &&
+                    cumulativeSpanContext is not null &&
+                    extractor.PropagatorType == PropagatorType.TraceContext)
                 {
-                    if (_propagationExtractFirstOnly)
-                    {
-                        return spanContext;
-                    }
+                    // skip this extractor if
+                    // - we are configured to extract only the first trace context (aka SpanContext)
+                    // - we already extracted trace context
+                    // - this extractor is a trace context extractor
+                    continue;
+                }
+
+                if (!extractor.TryExtract(carrier, carrierGetter, out var currentExtractedContext))
+                {
+                    // extractor failed to extract
+                    continue;
+                }
 
-                    if (localSpanContext is not null && spanContext is not null && _extractors[i] is W3CTraceContextPropagator)
+                if (cumulativeBaggage is null)
```

Comment:
Can we extract the logic to merge `cumulativeBaggage` and `currentExtractedContext.Baggage` into a separate small method? This loop in general is not the easiest to understand so if we can make the instructions here as succinct as possible, I think it will greatly help readability

Reply:
Agreed. I simplified this by always initializing `cumulativeBaggage = new()` so it's never null, and skip checking if the extracted baggage is null or empty, since that happens inside `Merge()` anyway.

Now it's just a single line (and I added a comment to explain why we accumulate all baggage):
```
cumulativeBaggage.Merge(currentExtractedContext.Baggage);
```